### PR TITLE
[SPARK-38823][SQL] Make `NewInstance` non-foldable to fix aggregation buffer corruption issue

### PR DESCRIPTION
**Pull request by bersprockets**
### What changes were proposed in this pull request?

Make `NewInstance` non-foldable.

### Why are the changes needed?

When handling Java beans as input, Spark creates `NewInstance` with no arguments. On master and 3.3, `NewInstance` with no arguments is considered foldable. As a result, the `ConstantFolding` rule converts `NewInstance` into a `Literal` holding an instance of the user's specified Java bean. The instance becomes a singleton that gets reused for each input record (although its fields get updated by `InitializeJavaBean`).

Because the instance gets reused, sometimes multiple buffers in `AggregationIterator` are actually referring to the same Java bean instance.

Take, for example, the test I added in this PR, or the `spark-shell` example I added to SPARK-38823 as a comment.

The input is:
```
    new Item("a", 1),
    new Item("b", 3),
    new Item("c", 2),
    new Item("a", 7)
```
As `ObjectAggregationIterator` reads the input, the buffers get set up as follows (note that the first field of Item should be the same as the key):
```
- Read Item("a", 1)

- Buffers are now:
  Key "a" --> Item("a", 1)

- Read Item("b", 3)

- Buffers are now:
  Key "a" -> Item("b", 3)
  Key "b" -> Item("b", 3)
```

The buffer for key "a" now contains `Item("b", 3)`. That's because both buffers contain a reference to the same Item instance, and that Item instance's fields were updated when `Item("b", 3)` was read.

This PR makes `NewInstance` non-foldable, so it will not get optimized away, thus ensuring a new instance of the Java bean for each input record.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 1]
```diff
@@ -516,6 +516,7 @@ case class NewInstance(
 
   override def nullable: Boolean = needNullCheck
 
+  override def foldable: Boolean = false
```
Comment (bersprockets) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:519]: Alternatively, I could make it false only when there are no arguments (i.e., `children.size` is zero).

**Comment by bersprockets** [comment]
ping @AngersZhuuuu @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:519]: Good catch! I think this change makes sense. `NewInstance` should create a fresh instance per invocation. Can we add some code comments here?

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Review by AngersZhuuuu — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 1 above)
Comment (bersprockets) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:519]: I will add a comment.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.3.

**Pull request closed without merging**